src/wasm/cpp/tests/wasm-cpp-test.ts

```diff
@@ -11,56 +11,220 @@ import {assert} from '../../../platform/chai-web.js';
-import {SingletonStorageProvider} from '../../../runtime/storage/storage-provider-base.js';
-import * as util from '../../../runtime/testing/test-util.js';
+import {RozSlotComposer} from '../../../runtime/testing/fake-slot-composer.js';
+import {VolatileSingleton, VolatileCollection} from '../../../runtime/storage/volatile-storage.js';
+import {assertThrowsAsync} from '../../../runtime/testing/test-util.js';
 
-describe('wasm C++ tests', () => {
+const schemasFile = 'src/wasm/cpp/tests/schemas.arcs';
+const wasmFile = 'build/wasm/cpp/tests/test-module.wasm';
+
+class TestLoader extends Loader {
+  resolve(path: string) {
+    return (path[0] === '$') ? `RESOLVED(${path})`: path;
+  }
+
+  clone(): TestLoader {
+    return this;
+  }
+}
+
+async function setup(manifestString) {
+  const loader = new TestLoader();
+  const manifest = await Manifest.parse(manifestString, {loader, fileName: process.cwd() + '/input.arcs'});
+  const runtime = new Runtime(loader, RozSlotComposer, manifest);
+  const arc = runtime.newArc('wasm-test', 'volatile://');
+
+  const recipe = arc.context.recipes[0];
+  recipe.normalize();
+  await arc.instantiate(recipe);
+  await arc.idle;
+
+  const [info] = arc.loadedParticleInfo.values();
+  return {arc, stores: info.stores, slotComposer: arc.pec.slotComposer as RozSlotComposer};
+}
+
+describe('wasm tests (C++)', () => {
   // TODO: https://github.com/PolymerLabs/arcs/issues/3418
   before(function() {
     if (!global['testFlags'].enableWasm) {
       this.skip();
     }
   });
 
-  it('simple entity passthrough', async () => {
-    const loader = new Loader();
-    const manifest = await Manifest.parse(`
-      import 'build/wasm/cpp/tests/test.arcs'
+  it('onHandleSync / onHandleUpdate', async () => {
+    const {arc, stores} = await setup(`
+      import '${schemasFile}'
+
+      particle HandleSyncUpdateTest in '${wasmFile}'
+        in Data input1
+        in Data input2
+        out [Data] output
+
       recipe
-        PassThrough
-          input <- h0
-          output -> h1
-      `, {loader, fileName: process.cwd() + '/input.arcs'});
+        HandleSyncUpdateTest
+          input1 <- h1
+          input2 <- h2
+          output -> h3
+      `);
+    const input1 = stores.get('input1') as VolatileSingleton;
+    const input2 = stores.get('input2') as VolatileSingleton;
+    const output = stores.get('output') as VolatileCollection;
+
+    // onHandleSync: txt = 'sync:<handle-name>'; flag = all_synced
+    // The order in which handles are synchronized isn't guaranteed, so allow for either result.
+    const syncs = (await output.toList()).map(e => e.rawData);
+    const expected = ['sync:input1', 'sync:input2'];
+    if (syncs[0].txt === expected[0]) {
+      assert.deepStrictEqual(syncs, [{txt: expected[0], flg: false}, {txt: expected[1], flg: true}]);
+    } else {
+      assert.deepStrictEqual(syncs, [{txt: expected[1], flg: false}, {txt: expected[0], flg: true}]);
+    }
```

Comment:
I believe this can all be replaced with:
```
assert.sameDeepMembers(syncs, [{txt: 'sync:input1', flg: false}, {txt: 'sync:input2', flg: true}]);
```

`sameDeepMembers` checks for all expected values (and no others), in any order.

Reply:
No, the problem isn't the ordering of the list, it's whether the `flg` value is true or false against input 1 or 2.

Reply:
Oooooohhhhh, I didn't notice that. In that case I've got no nicer suggestions :-(